sdk/rust/crates/dagger-sdk/src/core/cli_session.rs

```diff
@@ -19,27 +25,54 @@ impl CliSession {
     pub async fn connect(
         &self,
         config: &Config,
-        cli_path: &PathBuf,
-    ) -> eyre::Result<(ConnectParams, tokio::process::Child)> {
+        cli_path: &Path,
+    ) -> eyre::Result<(ConnectParams, DaggerSessionProc)> {
         self.inner.connect(config, cli_path).await
     }
 }
 
+pub struct DaggerSessionProc {
+    inner: tokio::process::Child,
+}
+
+impl Drop for DaggerSessionProc {
+    fn drop(&mut self) {
+        // FIXME: This is a bit of a hack, we spawn a neutral runtime, which doesn't automatically prune blocking tasks.
+        // This should be removed when async_drop is made stable
+        futures::executor::block_on(async move {
+            tracing::trace!("waiting for dagger subprocess to shutdown");
+
+            if let Err(e) = self.inner.wait().await {
```

Comment:
As explained to me in Discord, this `.wait()` also does the job of closing `stdin`:

https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.wait

> The stdin handle to the child process, if any, will be closed before waiting. This helps avoid deadlock: it ensures that the child does not block waiting for input from the parent, while the parent waits for the child to exit.